README.rst
g2tools
-------

This module contains a small number of tools useful for analyzing
contributions to the muon's magnetic moment from (lattice) QCD vacuum
polarization. The functions or classes include:

    **moments(G)**
        Compute moments of *jj* correlator *G*.

    **mom2taylor(mom)**
        Convert moments ``mom`` into Taylor coefficients for *q*\ :sup:`2`-expansion.

    **taylor2mom(tayl)**
        Convert Taylor expansion coefficients ``tayl`` into moments.

    **vacpol(mom)**
        Create a Pade approximant for the subtracted
        vacuum polarization (``PI-hat``) from the *jj* correlator
        whose moments (or Taylor coefficients) are in *mom*.

    **fourier_vacpol(G)**
        Create subtracted vacuum polarization (``PI-hat``) by
        Fourier transforming *jj* correlator ``G(t)``.

    **a_mu(pihat, Q)**
        Compute the contribution to the muon's *g-2*
        anomaly from function pihat (usually built by vacpol).

    **R2G(E, R)**
        Compute the Euclidean G(t) corresponding to data 
        for Re+e-.

    **R2a_mu(E, R)**
        Compute the leading-order contribution to the 
        muon's *g-2* anomaly corresponding to data 
        for Re+e-.

    **TanhWin(t0, t1, dt)**
        Create a filter for applying a t-window in 
        ``monents(...)`` or ``fourier_vacpol(...)``.

    **pade_gvar(f, m, n)**
        General-purpose code for determining Pade approximants
        to a power series whose coefficients are ``GVar``\s (ie,
        Gaussian random variables, for error propagation).

    **pade_svd(f, m, n)**
        General-purpose code for determining Pade approximants
        for a power series whose coefficients are floats.
        Uses *svd* regularization to stabilize results when
        the input data are noisy.

Information on how to install the module is in the file INSTALLATION.

To test the module try ``make tests``.

Documentation is in the doc directory: open doc/html/index.html
or look online at <https://g2tools.readthedocs.io>.

The examples directory has a complete example, showing how to go from Monte
Carlo data for a *jj* correlator to a contribution to the muon's magnetic
moment anomaly *a*\ :sub:`µ`. See also the introduction in the documentation.

The general technique that underlies this module is described in
Chakraborty *et al*, Phys.Rev. D89 (2014) no.11, 114501. Google
``arXiv:1403.1778`` to find a preprint on the web.

| Created by G. Peter Lepage (Cornell University) on on 2014-09-13.
| Copyright (c) 20014-22 G. Peter Lepage.

.. image:: https://zenodo.org/badge/66222496.svg
   :target: https://zenodo.org/badge/latestdoi/66222496



CHANGES
Version 1.4.4 2023-05-12
==========================
- Updates installation code.

Version 1.4.1  2022-04-16
===========================
Adds missing documentation (about windows) and fixes default dt in TanhWin.
Fixes bug in TanhWin when both t0 and t1 are None.


Version 1.4 2022-04-07
========================

- fourier_vacpol() and moments() have a new keyword, filter, that allows
  the use of t-windows, among other things.

- Adds class TanhWin which is a t-window filter for use in the above.

- Adds new methods R2G and R2a_mu for analyzing Re+e- data.

Version 1.3.4 2018-06-26
=========================

- moments() recoded to enforce periodicity when periodic=True (used to
  be assumed).

- fourier_vacpol() and moments() both have new keywords tmin and tmax
  to limit the part of correlator used in the vac. pol. function.

- New tests.

Version 1.3.3 2018-02-19
========================
Minor update. This has tiny improvements to the installation script (setup.py)
and distribution file layout to make  installation more robust, with less
opportunity for confusion on the  part of the user.

It also includes a piece of the documentation (for g2tools.a_mu) that
was missing before.

Version 1.3.2 2018-02-15
========================
Updated interface to scipy. Cleaned up installation minor issues.


Version 1.3 2017-06-13
======================

- Significant rewrite of pade_svd and especially pade_gvar to clean them up.
  Also the relative tolerance rtol is more flexible now.

- vacpol will now raise an exception if the poles in its
  Pade approximant are complex or too close to being positive. This
  behavior can be turned off, but that is likely to be a bad idea.

- vacpol has new method FT(t) which Fourier transforms the Pade back
  into t space.

- vacpol.vector restricts its Pade to order=(1,1) because that is exact.

- Bug fix in fourier_vacpol for periodic=True

Version 1.2.1 2017-05-26
========================

- Bug fix for unusual crashes of pade_svd.

- New option qth for vacpol.

- More tests.

- Default rtol for pade_gvar is now set equal to the geometric
  mean of the ratios f[i].sdev/abs(f[i].mean) for each of the
  Taylor coefficients f[i].

Version 1.2 2017-05-25
======================

- Bug fix so vacpol.taylor() works.

- New option warn in vacpol. Setting warn=True causes a warning to be
  issued if either badpoles() is True or if the order of the Pade is
  reduced.

Version 1.1 2017-05-23
======================

- New class fourier_vacpol that creates a subtracted vacuum polarization
  function from the Fourier transform of G(t). This provides an alternative
  to using moments of G(t).

- New argument qmin for a_mu.

- New method vacpol.taylor which returns Taylor coefficients
  (shorthand for vpol.pseries['taylor'].c).

- Small bug in vacpol.vector --- set n=20 no matter what n set to.


Version 1.0.1  13/03/2017
=========================
Tiny change to make g2tools compatible with the latest lsqfit (v9).



examples/amu-example.py
""" Example application of g2tools.py

This example illustrates the use of :mod:`g2tools` to convert
lattices data (adapted from Chakraborty et al, arXiv:1403.1778)
into a contribution to the muon's magnetic anomaly.
"""
from __future__ import print_function

import g2tools as g2
import gvar as gv

def main():
    # data
    Z = gv.gvar('0.9938(17)')           # current Z factor
    Q = 1. / 3.                         # charge of quark (units of proton charge)
    ainv = gv.gvar('1.6280(86)')        # inverse lattice spacing (in GeV)

    G = gv.gvar([                       # G(t) for t=0..63 (in lattice units)
        '0.0870904(11)', '0.0435138(14)', '0.00509859(48)', '0.00305614(43)',
        '0.00069516(19)', '0.00045466(15)', '0.000166972(80)', '0.000102219(58)',
        '0.000045284(34)', '0.000026213(22)', '0.000012630(14)', '7.0635(91)e-06',
        '3.5569(57)e-06', '1.9469(37)e-06', '1.0027(24)e-06', '5.421(16)e-07',
        '2.834(10)e-07', '1.5174(67)e-07', '7.943(43)e-08', '4.253(28)e-08',
        '2.221(19)e-08', '1.183(12)e-08', '6.132(81)e-09', '3.292(51)e-09',
        '1.727(34)e-09', '9.19(22)e-10', '4.81(14)e-10', '2.643(96)e-10',
        '1.385(64)e-10', '7.61(44)e-11', '3.92(31)e-11', '2.67(24)e-11',
        '2.07(21)e-11', '2.90(23)e-11', '4.12(31)e-11', '8.20(42)e-11',
        '1.380(65)e-10', '2.788(98)e-10', '5.01(15)e-10', '9.72(23)e-10',
        '1.782(34)e-09', '3.406(53)e-09', '6.333(78)e-09', '1.212(12)e-08',
        '2.249(18)e-08', '4.283(28)e-08', '8.016(44)e-08', '1.5263(67)e-07',
        '2.843(10)e-07', '5.420(16)e-07', '1.0062(25)e-06', '1.9453(39)e-06',
        '3.5611(58)e-06', '7.0675(93)e-06', '0.000012647(14)', '0.000026240(22)',
        '0.000045282(32)', '0.000102285(56)', '0.000166993(79)', '0.00045479(15)',
        '0.00069503(19)', '0.00305647(42)', '0.00509870(47)', '0.0435158(14)'
        ])
    # N.B.: In general would construct G so that correlations from one t
    #   to the next are included. Don't bother here since this is meant
    #   just to illustrate g2tools.

    # compute moments, converting to physical units from lattice units
    mom = g2.moments(G, ainv=ainv, Z=Z, periodic=True, nlist=[4, 6, 8, 10])
    print('Taylor coefficients:', g2.mom2taylor(mom))
    print()

    # construct subtracted vac pol function using [2,2] Pade
    vpol = g2.vacpol(mom, order=(2,2))

    # integrate vpol to get a_mu and print result
    a_mu = g2.a_mu(vpol, Q=Q)
    print('1) a_mu contribution =', a_mu)
    print()

    # error budget for a_mu
    print(gv.fmt_errorbudget(
        outputs=dict(a_mu=a_mu, mom4=mom[4]),
        inputs=dict(G=G, Z=Z, ainv=ainv),
        ))

    # do it again with exact vac. pol. instead of moments
    vpol = g2.fourier_vacpol(G, ainv=ainv, Z=Z, periodic=True)
     # integrate vpol to get a_mu and print result
    a_mu = g2.a_mu(vpol, Q=Q)
    print('2) a_mu contribution =', a_mu)
    print()

    # error budget for a_mu
    print(gv.fmt_errorbudget(
        outputs=dict(a_mu=a_mu, mom4=mom[4]),
        inputs=dict(G=G, Z=Z, ainv=ainv),
        ))

    # do it again with a window
    vpol = g2.fourier_vacpol(
        G, ainv=ainv, Z=Z, periodic=True, filter=g2.TanhWin(t1=1.5, dt=0.15)
        )
     # integrate vpol to get a_mu and print result
    a_mu = g2.a_mu(vpol, Q=Q)
    print('3) a_mu contribution (win) =', a_mu)
    print()

    # error budget for a_mu
    print(gv.fmt_errorbudget(
        outputs=dict(a_mu=a_mu, mom4=mom[4]),
        inputs=dict(G=G, Z=Z, ainv=ainv),
        ))


if __name__ == '__main__':
    main()


# Created by G. Peter Lepage (Cornell University) on 2016-08-20.
# Copyright (c) 2016-22 G. Peter Lepage.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version (see <http://www.gnu.org/licenses/>).
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
